UI/BuildDesignatorWnd.cpp

```diff
@@ -241,25 +244,109 @@ namespace {
+        else
+            return ConditionDescription(location_conditions, GetUniverseObject(candidate_object_id), source);
     }
 
     boost::shared_ptr<GG::BrowseInfoWnd> ProductionItemRowBrowseWnd(const ProductionQueue::ProductionItem& item,
                                                                     int candidate_object_id, int empire_id)
     {
+        std::string title;
+        std::string main_text;
+        float total_cost;
+        int production_time;
+
+        // production item is a building
         if (item.build_type == BT_BUILDING) {
+            const BuildingType* building_type = GetBuildingType(item.name);
+            if (!building_type) {
+                boost::shared_ptr<GG::BrowseInfoWnd> browse_wnd;
+                return browse_wnd;
+            }
+
+            // create title, description, production time and cost
+            title = UserString(item.name);
+            main_text = UserString(building_type->Description());
+            total_cost = building_type->ProductionCost(empire_id, candidate_object_id);
+            production_time = building_type->ProductionTime(empire_id, candidate_object_id);
+
+            main_text += "\n\n" + UserString("PRODUCTION_WND_TOOLTIP_PROD_COST") + ": " + DoubleToString(total_cost, 3, false);
+            main_text += "\n" + UserString("PRODUCTION_WND_TOOLTIP_PROD_TIME") + ": " + boost::lexical_cast<std::string>(production_time);
+
+            // show build conditions
+            std::string enqueue_and_location_condition_failed_text = EnqueueAndLocationConditionDescription(item.name, candidate_object_id, empire_id, true);
+            if (!enqueue_and_location_condition_failed_text.empty())
+                if (TemporaryPtr<UniverseObject> location = GetUniverseObject(candidate_object_id)) {
+                    std::string failed_cond_loc = boost::io::str(FlexibleFormat(UserString("PRODUCTION_WND_TOOLTIP_FAILED_COND")) % location->Name());
+                    main_text += "\n\n" + failed_cond_loc + ":\n" + enqueue_and_location_condition_failed_text;
+            }
+
+            // create tooltip
             boost::shared_ptr<GG::BrowseInfoWnd> browse_wnd(new IconTextBrowseWnd(
-                ClientUI::BuildingIcon(item.name), UserString(item.name),
-                EnqueueAndLocationConditionDescription(item.name, candidate_object_id, empire_id)));
+                ClientUI::BuildingIcon(item.name), title, main_text));
             return browse_wnd;
-        } else if (item.build_type == BT_SHIP) {
-            const ShipDesign* ship_design = GetShipDesign(item.design_id);
-            const std::string& name = (ship_design ? ship_design->Name() : UserString("SHIP_DESIGN"));
+        }
+
+        // production item is a ship
+        if (item.build_type == BT_SHIP) {
+            const ShipDesign* design = GetShipDesign(item.design_id);
+            if (!design) {
+                boost::shared_ptr<GG::BrowseInfoWnd> browse_wnd;
```

Comment:
Passing an invalid design ID doesn't produce a crash. The function exits as intended, no tooltip is shown.